usim/py/events.py

```diff
@@ -0,0 +1,642 @@
+
+
+class Initialize:
+    """
+    Internal initializer for :py:class:`~.Process`
+
+    Since this event is internal and not used by 'usim.py',
+    it is not implemented. Instantiating this class
+    unconditionally raises an error.
+    """
+    def __init__(self, *args, **kwargs):
+        raise NotEmulatedError(self.__doc__)
+
+
+class InterruptQueue:
+    """
+    Internal helper to manage interrupts for :py:class:`~.Process`
+    """
+    __slots__ = '__usimpy_flag__', '_causes', 'ok', 'defused'
+
+    def __init__(self):
+        self.__usimpy_flag__ = Flag()
+        self._causes = []
+        self.ok = False
+        self.defused = True
+
+    def __bool__(self):
+        return bool(self._causes)
+
+    @property
+    def value(self):
+        """The next :py:exc:`Interrupt`; this is a destructive action"""
+        return Interrupt(self.pop())
+
+    def push(self, cause):
+        """Add a new interrupt with ``cause``"""
+        self._causes.append(cause)
+        if not self.__usimpy_flag__:
+            self.__usimpy_flag__._value = True
+            self.__usimpy_flag__.__trigger__()
+
+    def pop(self):
+        result = self._causes.pop(0)
+        if not self._causes:
+            self.__usimpy_flag__._value = False
+        return result
+
+
+class Process(Event[V]):
+    """
+    A concurrently running process which triggers as an event on completion
+
+    Processes wrap around generators, which in turn may ``yield`` events to
+    wait for them. This implements cooperative multitasking, allowing another
+    process to run while others wait. Each :py:class:`~.Process` acts as an
+    event which is automatically triggered when the underlying generator completes.
+
+    .. code:: python3
+
+        def clock(env: Environment, sound):
+            for _ in range(60):
+                print(sound)
+                yield env.timeout(1)
+
+        def clocks(env: Environment):
+            env.process('tick')
+            env.process('tack')
+            yield env.process('TOCK')
+
+    .. hint::
+
+        **Migrating to μSim**
+
+        When migrating from :py:class:`~.Process`, activities are expressed as
+        ``async def`` coroutines instead of generators. Activities do not need
+        an environment - it is implicitly available. An activity can ``await``
+        a :term:`notification`, concurrently iterate using ``async for``
+        and manage resources using ``async with``.
+
+        .. code:: python3
+
+            async def clock(sound):
+                for _ in range(60):
+                    print(sound)
+                    await (time + 1)
+
+        To run an activity, one must distinguish between a concurrent
+        :py:class:`~usim.typing.Task` and a nested activity. When an activity
+        exclusively waits for another activity, it can ``await`` it.
+
+        .. code:: python3
+
+            async def double_clock(sound):
+                "A clock of double duration"
```

Comment:
This cannot be copied to try out the example. You could use a common comment here, for example.

Reply:
```suggestion
                '''A clock of double duration'''
```